### ISSUE TITLE
OpenID connect with Azure AD breaks due to email not being present in response

### ISSUE THREAD

**Issue by SneWs** [uxbox/uxbox#1460]
When using OpenID to connect to a tennant in Azure AD the code needs to validate if the email field is null, and if so, check if preferred_username is an email address, and use that. Another option would be to require Azure AD applications to include one of the following properties:
 * http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress
 * verified_primary_email

Another feasible alternative might be to use oidc/unique_name as email address if email is null

**To Reproduce**
1. Create a new application in Azure AD and enable it to be used for sign in
2. Configure PENPOT_OIDC_* with the information needed for the client id and secret etc
3. Define PENPOT_OIDC_SCOPES to be of value "openid User.Read profile"

**Expected behavior**
Penpot should not fail auth flows for Azure AD OpenID sign in

**Environment (please complete the following information):**
This is a self hosted instance running the official docker containers

If self-hosted instance, add OS and runtime information to help explain your problem.
Official Docker containers

```
penpot-backend_1   | [2022-01-11 13:44:37.245] [clojure-agent-send-off-pool-6] W app.http.oauth - hint="received incomplete profile info object (please set correct scopes)" info="{:backend "oidc", :email nil, :fullname "REDACTED", :props {:oidc/sub "REDACTED", :oidc/uti "REDACTED", :oidc/amr "[\"pwd\"]", :oidc/unique_name "REDACTED@REDACTED.COM", :oidc/aio "E2ZgYOD59fdcQtXDvYrM98xcFSxjF2wLLT10lOPjbpcy5cNsUxQA", :oidc/oid "REDACTED", :oidc/upn "REDACTED", :oidc/ipaddr "188.149.204.44", :oidc/rh "0.AUcAj3Rjj2tI0E-dp5ZAVe5ne0IasQqxGrJKtTVNA09GNjlHAPI.", :oidc/family_name "REDACTED", :oidc/given_name "REDACTED", :oidc/tid "8f63748f-486b-4fd0-9da7-964055ee677b", :oidc/ver "1.0"}}"

penpot-backend_1   | [2022-01-11 13:44:37.256] [clojure-agent-send-off-pool-5] W app.http.oauth - hint="error on oauth process"
penpot-backend_1   | clojure.lang.ExceptionInfo: no user info
penpot-backend_1   |    at app.common.exceptions$error.invokeStatic(exceptions.cljc:34) ~[?:?]
penpot-backend_1   |    at app.common.exceptions$error.doInvoke(exceptions.cljc:26) ~[?:?]
penpot-backend_1   |    at clojure.lang.RestFn.invoke(RestFn.java:512) ~[clojure-1.10.3.jar:?]
penpot-backend_1   |    at app.http.oauth$retrieve_info.invokeStatic(oauth.clj:113) ~[?:?]
penpot-backend_1   |    at app.http.oauth$retrieve_info.invoke(oauth.clj:102) ~[?:?]
penpot-backend_1   |    at app.http.oauth$callback_handler.invokeStatic(oauth.clj:231) ~[?:?]
penpot-backend_1   |    at app.http.oauth$callback_handler.invoke(oauth.clj:228) ~[?:?]
penpot-backend_1   |    at app.http.oauth$wrap_handler$fn__29968.invoke(oauth.clj:261) ~[?:?]
penpot-backend_1   |    at ring.middleware.cookies$wrap_cookies$fn__32168.invoke(cookies.clj:214) ~[?:?]
penpot-backend_1   |    at app.http.middleware$wrap_errors$fn__32403.invoke(middleware.clj:106) ~[?:?]
penpot-backend_1   |    at app.http.middleware$wrap_parse_request_body$fn__32393.invoke(middleware.clj:55) ~[?:?]
penpot-backend_1   |    at ring.middleware.keyword_params$wrap_keyword_params$fn__32214.invoke(keyword_params.clj:53) ~[?:?]
penpot-backend_1   |    at ring.middleware.multipart_params$wrap_multipart_params$fn__32346.invoke(multipart_params.clj:171) ~[?:?]
penpot-backend_1   |    at ring.middleware.params$wrap_params$fn__32370.invoke(params.clj:67) ~[?:?]
penpot-backend_1   |    at app.http.middleware$wrap_format_response_body$fn__32399.invoke(middleware.clj:94) ~[?:?]
penpot-backend_1   |    at app.http.middleware$wrap_etag$fn__32417.invoke(middleware.clj:146) ~[?:?]
penpot-backend_1   |    at reitit.ring$ring_handler$fn__34243.invoke(ring.cljc:329) ~[?:?]
penpot-backend_1   |    at app.http.middleware$wrap_server_timing$fn__32380.invoke(middleware.clj:26) ~[?:?]
penpot-backend_1   |    at clojure.lang.AFn.applyToHelper(AFn.java:154) ~[clojure-1.10.3.jar:?]
penpot-backend_1   |    at clojure.lang.AFn.applyTo(AFn.java:144) ~[clojure-1.10.3.jar:?]
penpot-backend_1   |    at clojure.lang.AFunction$1.doInvoke(AFunction.java:31) ~[clojure-1.10.3.jar:?]
penpot-backend_1   |    at clojure.lang.RestFn.invoke(RestFn.java:408) ~[clojure-1.10.3.jar:?]
penpot-backend_1   |    at app.http$router_handler$fn__34895.invoke(http.clj:92) ~[?:?]
penpot-backend_1   |    at ring.adapter.jetty9$proxy_handler$fn__34788.invoke(jetty9.clj:54) ~[?:?]
penpot-backend_1   |    at ring.adapter.jetty9.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source) ~[?:?]
penpot-backend_1   |    at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:51) ~[jetty-server-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:171) ~[jetty-server-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122) ~[jetty-server-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.server.Server.handle(Server.java:562) ~[jetty-server-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.server.HttpChannel.lambda$handle$0(HttpChannel.java:406) ~[jetty-server-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:663) ~[jetty-server-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:398) ~[jetty-server-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:282) ~[jetty-server-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:319) ~[jetty-io-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100) ~[jetty-io-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.io.SocketChannelEndPoint$1.run(SocketChannelEndPoint.java:101) ~[jetty-io-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:412) ~[jetty-util-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:381) ~[jetty-util-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:268) ~[jetty-util-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.lambda$new$0(AdaptiveExecutionStrategy.java:138) ~[jetty-util-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:378) ~[jetty-util-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:894) ~[jetty-util-10.0.6.jar:10.0.6]
penpot-backend_1   |    at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1038) ~[jetty-util-10.0.6.jar:10.0.6]
penpot-backend_1   |    at java.lang.Thread.run(Thread.java:831) [?:?]
```

** Additions **
This SO thread also talks about this issue https://stackoverflow.com/questions/30983694/get-the-users-email-address-from-azure-ad-via-openid-connect and it also links to https://docs.microsoft.com/en-gb/azure/active-directory/develop/v2-permissions-and-consent#openid-connect-scopes that clarifies that the email property will not always be present for Azure AD. However, if you where to use the consumer version of MSA the email field will be populated

**Comment by niwinz** [comment]
Hello @SneWs
I'm not clearly understand the problem. Penpot requires that email field is present on the openid response. Each openid provider is different and implements the scopes as they want. A compliant openid provider should be able to implement the `email` scope that will expose `email` and `email_verified` (https://openid.net/specs/openid-connect-basic-1_0.html#Scopes)

I think implementing peculiarities for each openid provider is out of scope of penpot.

There are no way in Azure AD expose properly the email property?

**Comment by SneWs** [comment]
No, that is correct, Azure AD does not expose the email field, and from what I can tell, there is no way to force it, their SDKs and helpers use the other properties as I mentioned in the initial post to grab the email. And the OpenID spec does not require the email property, it's marked as optional, so Microsoft is not failing to follow the specification here to my knowledge.

I fully understand that it might not be great to have to add an additional branch to check this, but Azure AD is kind of widly used and the other option would be to add a completely separate config section and handling for Microsoft Authentication, but since they do support OpenID protocol the additional check feels to me as the simple way to solve this.

**Comment by niwinz** [comment]
I understand.

Then, what is the less error prone attribute that can be treat as email on Azure AD. I think we can make it a little bit flexible adding an additional setting for it. Something like: `PENPOT_OIDC_EMAIL_PROPS="email,other_field".

I have one question, is `http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress` can be a field key on the azure response?

**Comment by niwinz** [comment]
In any case, after reading the SO response, looks like the microsoft openid is not so compliant.

> Note that the e-mail address may not be returned in an email claim: in my case (once I got it working) it's coming back in a name claim.

> As per this guide to Scopes, permissions, and consent in the Azure Active Directory v2.0 endpoint, even if you include the email scope you may not get an e-mail address back...

Also reading the MS, documentation, it looks very clear, it is compliant but has the case where no email is associated with the account.

> The email scope can be used with the openid scope and any other scopes. It gives the app access to the user's primary email address in the form of the email claim.

> The email claim is included in a token only if an email address is associated with the user account, which isn't always the case. If your app uses the email scope, the app needs to be able to handle a case in which no email claim exists in the token.

So, in this case, when no email is associated with account, it is not possible to create an account on penpot because we use email as primary key for profile lookup. I think in any case that the best and the correct solution is properly assign a correct email on the Azure AD to the accounts.

As a future workaround for this case, we can think on a good solution from penpot side, allowing profiles without email maybe?

In summary, if the account on Azure AD has email associated it should come correctly under `email` property as MS doc says. And if the email is null, this is probably no other fields will contain email address; so then, I think implementing a quick & dirty workaround for this case is not worth...

I'm wrong?

**Comment by SneWs** [comment]
I don't think any one of us are wrong here, it's just what will work in a convinient and flexible enough way.

Based on the OpenID spec alone, Microsoft is conformant with it as it does not require the email field to be present, and for the corporate Azure AD, even if you have an email address set, it will never be exposed as the emil field oddly enough, and I have not found any possible way to force that in Azure AD.

I really like the idea of exposing config properties such as `PENPOT_OIDC_EMAIL_PROPS="email,other_field` 
Perhaps even have such overrride configs for each required parameter for the create account/signup flow in OpenID and other external account providers, that way me and others would be able to adjust the result to the expected one in the code base

And I'm no big supporter of Microsoft tech in general, but based on the OpenID spec you linked, they are conforming to it as the email property is marked as optional in their specification.

I would love if we could make some progress on this as I've set up a selfhost of this to replace XD and our UI and UX artists are really liking the software, as do I. It's a great project, well done.

**Comment by SneWs** [comment]
And I missed responding to `http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress`
This is the full name of the property to get the email address out when i grab for it in our backend systems, so just using emailaddress will fail, the fully qualified key needs to be used unfortunatly

**Comment by niwinz** [comment]
> I would love if we could make some progress on this as I've set up a selfhost of this to replace XD and our UI and UX artists are really liking the software, as do I. It's a great project, well done.

Agree, 

I'm just thinking on how it can be solved in the best way without introducing and ad-hoc quick and dirty patch for solving a very specific case.

> And I'm no big supporter of Microsoft tech in general, but based on the OpenID spec you linked, they are conforming to it as the email property is marked as optional in their specification.

As far as I understand, the implementation of scope `email` is optional on the spec but if you implement it, you should be able to access the email. And here is when MS fails, it implements the scope `email` but does not guaranties email field. The correct solution would be provide a MS specific scope for retrieve an "Maybe<Email>"

> And I missed responding to http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress ...

Do you know when I can find documentation about this scope, and what props/claims it exposes on the userinfo endpoint?

Following on the main issue and reading the https://stackoverflow.com/a/49679604 response and ms doc, they suggest using the `preferred_username`, but I din't found it on the sample response from the error message you posted and didn't seen anywhere on the MS docs on how to access to it from userinfo endpoint :(

As far as I can understand from here https://docs.microsoft.com/en-gb/azure/active-directory/develop/userinfo#consider-use-an-id-token-instead, MS just has this for the mark "Yeah I'm implement openid connect" (but please dont use this endpoint, it is pretty useless hehe). They suggest don't look at the userinfo endpoint and just read the token; this is more problematic to change right now in a reasonable amount of time... And on top of it all, they return the unique_name claim on userinfo endpoint when it is [not defined on the OIDC standard](https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims) and they clearly says that nonstandard claims can not be added to userinfo :S   

I will think a bit more on it...

**Comment by SneWs** [comment]
`preferred_username` seems to be what dotnet core libraries uses when/if email property is null or empty, so that def. would be the way to go as that is always part of the response, even with a default configured Azure AD application for user authentication.

I dumped the claims I get back using the same Azure AD application as I did for Penpot, and the following claims are returned to me

```
aio=REDACTED
http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress=my-email@company.com
http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname=REDACTED
http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname=Marcus
name=Marcus
http://schemas.microsoft.com/identity/claims/objectidentifier=REDACTED
preferred_username=my-email@company.com
rh=REDACTED
http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier=REDACTED
http://schemas.microsoft.com/identity/claims/tenantid=REDACTED
uti=REDACTED
verified_primary_email=my-email@company.com
aio=REDACTED
http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress=my-email@company.com
```

The docs from Microsoft seems to be a bit lacking on the subject of OpenID, I've been trying to find something concrete and to the point, but all I end up finding is https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent and that is really not helpful.

IMHO, I'd say being able to configure what property is read from the response for email would be the more flexible option, and it can default to the current value of email thus keeping it backwards compatible and working the same as now, with the added benefit that me and others will be able to override it from config.

**Comment by niwinz** [comment]
> I dumped the claims I get back using the same Azure AD application as I did for Penpot, and the following claims are returned to me

but this claims looks like not coming in the userinfo endpoint, they are probably part of the token id, and this claims are not read by penpot impl.

**Comment by SneWs** [comment]
Ah, yes, you are correct, those values comes from the token. I'm not all that familiar with Java, the JVM or Closure, so excuse me here if I'm totally wrong.

But looking at the java examples, and assuming you also have full access to java based libraries and 3rd party packages, would it be a bad idea to use  `io.jsonwebtoken` for example and parse the token? Could be added as a fallback for when email is empty in combination with a config flag for example? 

Microsoft example code that I found https://github.com/Azure-Samples/ms-identity-java-webapp/blob/master/msal-java-webapp-sample/src/main/java/com/microsoft/azure/msalwebsample/AuthHelper.java

**Comment by niwinz** [comment]
Yeah, it is  not imposible, is just what right now it implies too much change. It will require leave my current priorities, setup test enviroment and develop this. It is not an "easy fix". I will think on it and get a decision soon.

**Comment by SneWs** [comment]
That is fully understandable. 
I really appreciate you taking the time with this. Thank you. 

And feel free to ping me if something comes up in regards to this.

**Comment by niwinz** [comment]
just FYI, i will try to prioritize this issue for the next release

**Comment by niwinz** [comment]
This is addressed in https://github.com/penpot/penpot/pull/1656 and probably will  be available on 1.13
On the end, I opted for a simple solution to add `PENPOT_OIDC_EMAIL_ATTR` and `PENPOT_OIDC_NAME_ATTR`
